Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py

```diff
@@ -139,7 +152,18 @@ def process(stage, message):
     # 'data_format' list contains one of the allowed formats
     # or not set at all.
     if update or not formats:
-        amiPhysValues(data)
+        try:
+            amiPhysValues(data)
+        except DataflowException:
+            raise
+        except SSLError:
+            stage.output_error("Failed to process dataset '%s'"
+                               % data['datasetname'], sys.exc_info())
+        except Exception:
+            e_type, e_value, e_trace = sys.exc_info()
+            sys.stderr.write("(ERROR) Failed to process dataset '%s': %s\n"
+                             % (data['datasetname'], e_value))
```

Comment:
Why handle `SSLError`s and other `Exception`s differently?
As far as I can understand, the difference here is in traceback -- which is provided for `SSLError`, but not for other exceptions.

I couldn't invent a regular way to get to this point, so for testing I hardcoded invalid parameters to be passed to `ami_client.execute(...)` ([here](https://github.com/PanDAWMS/dkb/blob/933747d6eb6f80be59e2342949739dbbe1e85a20/Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py#L185)), and get the following log:
```
[dkb@aiatlas171 095_datasetInfoAMI]$ vi amiDatasets.py
# Alter ami_client.execute() parameters...
[dkb@aiatlas171 095_datasetInfoAMI]$ echo '{"datasetname": "ttt"}' | ./amiDatasets.py -m s
<...>
2020-05-21 15:48:26 (INFO) (ProcessorStage) Starting stage execution.
(ERROR) Failed to process dataset 'ttt': 'Result'
{"datasetname": "ttt"}
2020-05-21 15:48:26 (INFO) (ProcessorStage) Stopping stage.
[dkb@aiatlas171 095_datasetInfoAMI]$ vi amiDatasets.py
# Alter again...
[dkb@aiatlas171 095_datasetInfoAMI]$ echo '{"datasetname": "ttt"}' | ./amiDatasets.py -m s
<...>
2020-05-21 15:48:58 (INFO) (ProcessorStage) Starting stage execution.
(ERROR) Failed to process dataset 'ttt': No JSON object could be decoded
{"datasetname": "ttt"}
2020-05-21 15:48:58 (INFO) (ProcessorStage) Stopping stage.
```

Below I provide the traceback that (could) be output in these cases (and values of the variables that provoked the exceptions):
(1)
```
2020-05-21 15:51:27 (DEBUG) (ProcessorStage) Traceback (most recent call last):
(==)   File "./amiDatasets.py", line 157, in process
(==)     amiPhysValues(data)
(==)   File "./amiDatasets.py", line 192, in amiPhysValues
(==)     rowset = json_str['AMIMessage'][0]['Result'][0]['rowset']
(==) KeyError: 'Result'
```
`json_str` variable value:
```
{"AMIMessage" : [{"error" : [{"$" :"Command not found `AMIGetPhysicsParamsForDatase`."}],"executionTime" : [{"$" :"1.590068902701E12 "}]}]}
```

(2)
```
2020-05-21 15:53:49 (DEBUG) (ProcessorStage) Traceback (most recent call last):
(==)   File "./amiDatasets.py", line 157, in process
(==)     amiPhysValues(data)
(==)   File "./amiDatasets.py", line 191, in amiPhysValues
(==)     json_str = json.loads(res)
(==)   File "/usr/lib64/python2.7/json/__init__.py", line 338, in loads
(==)     return _default_decoder.decode(s)
(==)   File "/usr/lib64/python2.7/json/decoder.py", line 366, in decode
(==)     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
(==)   File "/usr/lib64/python2.7/json/decoder.py", line 384, in raw_decode
(==)     raise ValueError("No JSON object could be decoded")
(==) ValueError: No JSON object could be decoded
```
`res` variable value:
```
u'AMI#\nerror   : Command not found `AMIGetPhysicsParamsForDatase`.\n#'
```

Reply:
> Why handle `SSLError`s and other `Exception`s differently?
> As far as I can understand, the difference here is in traceback -- which is provided for `SSLError`, but not for other exceptions.

`SSLError` was important because it allowed us to catch problems with authentication. Now, when a test command is performed before processing, it's not that important, so a special handling can be removed. On the other hand, normal `Exception`s may need more information: as can be seen in your example, `(ERROR) Failed to process dataset 'ttt': 'Result'` is not very informative.

Removed `SSLError` and moved its handling to `Exception`, to provide traceback and use `stage.output_error()`. Also changed the way said lack of `'Result'` is treated to display the (rightfully) suggested `json_str`.

Note - all these things really call for rebasing. Why add `SSLError` in one commit and remove it later in the other, in the same PR? And some other questionable points.
However, the rebase will be pretty deep, so I decided to do it later, when we'll conclude the discussion of this point.

> (2)
> 
> ```
> 2020-05-21 15:53:49 (DEBUG) (ProcessorStage) Traceback (most recent call last):
> (==)   File "./amiDatasets.py", line 157, in process
> (==)     amiPhysValues(data)
> (==)   File "./amiDatasets.py", line 191, in amiPhysValues
> (==)     json_str = json.loads(res)
> (==)   File "/usr/lib64/python2.7/json/__init__.py", line 338, in loads
> (==)     return _default_decoder.decode(s)
> (==)   File "/usr/lib64/python2.7/json/decoder.py", line 366, in decode
> (==)     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
> (==)   File "/usr/lib64/python2.7/json/decoder.py", line 384, in raw_decode
> (==)     raise ValueError("No JSON object could be decoded")
> (==) ValueError: No JSON object could be decoded
> ```
> 
> `res` variable value:
> 
> ```
> u'AMI#\nerror   : Command not found `AMIGetPhysicsParamsForDatase`.\n#'
> ```

This can also be caught, by checking whether the JSON was decoded correctly, but... You got this by changing the requested format in `res = ami_client.execute(...)`, right? If yes, then such check does not look necessary - we already check that AMI client works (by running test command), so we can expect its further responses to be coherent (in JSON if we request JSON). Unless something wrong happens in AMI after the test command, but before the request about a dataset, which looks like a far-fetched situation to me.

Reply:
> Note - all these things (as well as some other ones that were done/discussed previously) really call for rebasing. Why add SSLError in one commit and remove it later in the other, in the same PR? And some other questionable points.
> However, the rebase will be pretty deep, so I decided to do it later, when we'll conclude the discussion of this point.

OK.

>> (2)
>> ```
>> 2020-05-21 15:53:49 (DEBUG) (ProcessorStage) Traceback (most recent call last):
>> <...>
>> (==) ValueError: No JSON object could be decoded
>> ```
>> `res` variable value:
>> ```
>> u'AMI#\nerror   : Command not found `AMIGetPhysicsParamsForDatase`.\n#'
>> ```
>
> This can also be caught, by checking whether the JSON was decoded correctly, but... <...> such check does not look necessary - we already check that AMI client works (by running test command), so we can expect its further responses to be coherent (in JSON if we request JSON). Unless something wrong happens in AMI after the test command, but before the request about a dataset, which looks like a far-fetched situation to me.

Right, the provided example situation was just to illustrate how some random error handling is gonna look like. As for the response format -- well... even if the AMI server sends a valid JSON document, I guess on its way to our client it still may be broken by some misfunction of the network, local resources etc. So no, I would not bet that we are guaranteed to get a valid JSON even if we asked for one and have previously received one.

I am not saying, however, that this situation must be specially handled here -- it is still a (supposedly) rare and accidental situation, so the traceback output from the outermost exceptions handling of the `ProcessorStage` object should be enough to say that we got something unexpected in the response. Should it become a regular problem one day -- then it will require a more accurate handling, but now it is just as optional as an accurate handling of dozens of other possible accidents.

Reply:
> > > (2)
> > > ```
> > > 2020-05-21 15:53:49 (DEBUG) (ProcessorStage) Traceback (most recent call last):
> > > <...>
> > > (==) ValueError: No JSON object could be decoded
> > > ```
> > > 
> > > 
> > > `res` variable value:
> > > ```
> > > u'AMI#\nerror   : Command not found `AMIGetPhysicsParamsForDatase`.\n#'
> > > ```
> > 
> > 
> > This can also be caught, by checking whether the JSON was decoded correctly, but... <...> such check does not look necessary - we already check that AMI client works (by running test command), so we can expect its further responses to be coherent (in JSON if we request JSON). Unless something wrong happens in AMI after the test command, but before the request about a dataset, which looks like a far-fetched situation to me.
> 
> Right, the provided example situation was just to illustrate how some random error handling is gonna look like. As for the response format -- well... even if the AMI server sends a valid JSON document, I guess on its way to our client it still may be broken by some misfunction of the network, local resources etc. So no, I would not bet that we are guaranteed to get a valid JSON even if we asked for one and have previously received one.
> 
> I am not saying, however, that this situation must be specially handled here -- it is still a (supposedly) rare and accidental situation, so the traceback output from the outermost exceptions handling of the `ProcessorStage` object should be enough to say that we got something unexpected in the response. Should it become a regular problem one day -- then it will require a more accurate handling, but now it is just as optional as an accurate handling of dozens of other possible accidents.

Fair points. We leave it be for now then.